cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,372 @@
+            continue
+
+        if cnots[j][y] in qubit_to_index:
+            k = qubit_to_index[cnots[j][y]]
+            # The controls (resp. targets) are the same, so we can simplify away.
+            cnots = [cnots[n] for n in range(len(cnots)) if n != j and n != k]
+            return True, cnots
+
+        qubit_to_index[cnots[j][y]] = j
+
+    return False, cnots
+
+
+def _simplify_cnots_tripplets(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Simplifies CNOT pairs according to equations 11 of [4].
+
+    CNOT(i, j) @ CNOT(j, k) == CNOT(j, k) @ CNOT(i, k) @ CNOT(i, j)
+    ───@───────       ───────@───@───
+       │                     │   │
+    ───X───@───   =   ───@───┼───X───
+           │             │   │
+    ───────X───       ───X───X───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
+    """
+    x, y = (0, 1) if flip_control_and_target else (1, 0)
+
+    # We investigate potential pivots sequentially.
+    for j in range(1, len(cnots) - 1):
+        # First, we look back for as long as the targets (resp. controls) are the same.
+        # They all commute, so all are potential candidates for being simplified.
+        common_A: Dict[int, int] = {}
+        for i in range(j - 1, -1, -1):
+            if cnots[i][y] != cnots[j][y]:
+                break
+            # We take a note of the control (resp. target).
+            common_A[cnots[i][x]] = i
+
+        # Next, we look forward for as long as the controls (resp. targets) are the
+        # same. They all commute, so all are potential candidates for being simplified.
+        common_B: Dict[int, int] = {}
+        for k in range(j + 1, len(cnots)):
+            if cnots[j][x] != cnots[k][x]:
+                break
+            # We take a note of the target (resp. control).
+            common_B[cnots[k][y]] = k
+
+        # Among all the candidates, find if they have a match.
+        keys = common_A.keys() & common_B.keys()
+        for key in keys:
+            assert common_A[key] != common_B[key]
+            # We perform the swap which removes the pivot.
+            new_idx: List[int] = (
+                [idx for idx in range(0, j) if idx != common_A[key]]
+                + [common_B[key], common_A[key]]
+                + [idx for idx in range(j + 1, len(cnots)) if idx != common_B[key]]
+            )
+            # Since we removed the pivot, the length should be one fewer.
+            assert len(new_idx) == len(cnots) - 1
+            cnots = [cnots[idx] for idx in new_idx]
+            return True, cnots
+
+    return False, cnots
+
+
+def _simplify_cnots(cnots: List[Tuple[int, int]]):
+    """
```

Comment:
Nit no new line - the first line should be the short description of the function.

Reply:
Done.